### DISCUSSION TITLE
104. InnoDB에서 데드락이 발생하는 상황 하나를 설명하고 해결 방법까지 말해주세요.

### DISCUSSION THREAD

**Discussion by jongmee** [jewelry-box/cs-interview#108] [General]
키워드: 레코드 잠금이 데드락을 유발하는 상황, 데드락

**Comment by cutehumanS2** [comment]
InnoDB에서는 **트랜잭션 간 잠금 순서가 충돌**하면 데드락이 발생할 수 있습니다. 

예를 들어, 두 개의 트랜잭션이 동일한 두 테이블의 레코드를 서로 다른 순서로 잠그려고 할 때 두 트랜잭션이 서로 상대방의 잠금 해제를 기다리며 데드락이 발생합니다. 

- **트랜잭션 A**: 먼저 **Record X**를 잠근 후 **Record Y**를 잠그려고 시도.
- **트랜잭션 B**: 먼저 **Record Y**를 잠근 후 **Record X**를 잠그려고 시도.

<br>

데드락을 **방지**하기 위해 다음과 같은 방법을 고려해 볼 수 있습니다.
- 모든 트랜잭션이 항상 동일한 순서로 자원을 잠금하도록 설계합니다.
- 트랜잭션을 작고 짧게 유지하여 잠금 충돌 가능성을 줄입니다.
- 적절한 인덱스를 추가하여 쿼리의 스캔 범위를 줄이면 잠금 충돌 가능성을 줄일 수 있습니다.
- 불필요한 잠금 사용을 최소화합니다.
    - 가능하면 for update 또는 for share를 사용하지 않습니다.
    - 격리 수준을 낮추는 것도 방법 중 하나입니다.
- 테이블 수준 잠금 또는 세마포어 테이블 생성으로 트랜잭션을 직렬화합니다.

<br>

데드락을 **해결**하기 위해서는 다음과 같은 방법을 고려해 볼 수 있습니다.
- InnoDB가 데드락을 감지하여 트랜잭션 중 하나를 롤백하면, 자동으로 **재시도하는 로직**을 구현합니다.
- [SHOW ENGINE INNODB STATUS](https://dev.mysql.com/doc/refman/8.0/en/show-engine.html) 명령어로 최근 데드락 원인을 파악합니다.
    - [innodb_print_all_deadlocks](https://dev.mysql.com/doc/refman/8.0/en/innodb-parameters.html#sysvar_innodb_print_all_deadlocks)를 활성화하여 MySQL 오류 로그에 기록된 데드락 정보를 검토할 수도 있습니다.

<br><br>
[참고한 자료]
https://dev.mysql.com/doc/refman/8.0/en/innodb-deadlocks-handling.html
